File: src/openjarvis/tools/think.py
"""Think tool — zero-cost reasoning scratchpad."""

from __future__ import annotations

from typing import Any

from openjarvis.core.registry import ToolRegistry
from openjarvis.core.types import ToolResult
from openjarvis.tools._stubs import BaseTool, ToolSpec


@ToolRegistry.register("think")
class ThinkTool(BaseTool):
    """Reasoning scratchpad that echoes input for chain-of-thought."""

    tool_id = "think"

    @property
    def spec(self) -> ToolSpec:
        return ToolSpec(
            name="think",
            description=(
                "A reasoning scratchpad. Think through"
                " a problem step by step. Input is echoed."
            ),
            parameters={
                "type": "object",
                "properties": {
                    "thought": {
                        "type": "string",
                        "description": "Your reasoning or thought process.",
                    },
                },
                "required": ["thought"],
            },
            category="reasoning",
            cost_estimate=0.0,
            latency_estimate=0.0,
        )

    def execute(self, **params: Any) -> ToolResult:
        thought = params.get("thought", "")
        from openjarvis._rust_bridge import get_rust_module
        _rust = get_rust_module()
        content = _rust.ThinkTool().execute(thought)
        return ToolResult(
            tool_name="think",
            content=content,
            success=True,
        )


__all__ = ["ThinkTool"]

File: repos/openjarvis/rust/crates/openjarvis-tools/src/builtin/think.rs
//! Think tool — allows the agent to express reasoning steps.

use crate::traits::BaseTool;
use openjarvis_core::{OpenJarvisError, ToolResult, ToolSpec};
use once_cell::sync::Lazy;
use serde_json::Value;
use std::collections::HashMap;

static SPEC: Lazy<ToolSpec> = Lazy::new(|| ToolSpec {
    name: "think".into(),
    description: "Express a reasoning step without side effects".into(),
    parameters: serde_json::json!({
        "type": "object",
        "properties": {
            "thought": {
                "type": "string",
                "description": "The reasoning or thinking step"
            }
        },
        "required": ["thought"]
    }),
    category: "reasoning".into(),
    cost_estimate: 0.0,
    latency_estimate: 0.0,
    requires_confirmation: false,
    timeout_seconds: 1.0,
    required_capabilities: vec![],
    metadata: HashMap::new(),
});

pub struct ThinkTool;

impl BaseTool for ThinkTool {
    fn tool_id(&self) -> &str {
        "think"
    }

    fn spec(&self) -> &ToolSpec {
        &SPEC
    }

    fn execute(&self, params: &Value) -> Result<ToolResult, OpenJarvisError> {
        let thought = params["thought"].as_str().unwrap_or("(empty thought)");
        Ok(ToolResult::success("think", thought))
    }
}
